RFC-0007-plan-constraints-grammar.md

```diff
@@ -0,0 +1,77 @@
+# **RFC007 for Presto**
+
+## Grammar for Plan Constraints
+
+Proposers
+
+* @aaneja
+* @ClarenceThreepwood
+
+## Related Issues
+
+* [Overview doc](https://prestodb.io/wp-content/uploads/Search-Space-Improvements-Plan-Constraints.pdf) on Plan Constraints as a tool to control search space
+* PrestoDB blog on - [Elevating Presto Query Optimization](https://prestodb.io/blog/2024/03/21/elevating-presto-query-optimization/)
+
+## Summary
+
+This document proposes a grammar for specifying plan constraints
+
+## Background
+
+Plan constraints can be used to lock down critical aspects of an execution plan, such as access method, join method, and join order. 
+
+
+## Proposed Implementation
+
+We propose a grammar for specifying independent plan constraints, which take the form of a SQL comment block that would build an object graph of the constraints. Multiple constraints can be specified in a single place. The grammar is open for extension as we develop more mechanisms to lock down plans.
+
+In this first cut, users can build constraints to control 
+- Join orders and distributions for INNER JOIN's
+- Cardinality (row counts) for base relations and join sub-plans
+
+### Grammar
+
+```
+planConstraintString : /*! planConstraint [, ...] */
+
+planConstraint : joinConstraint
+| cardinalityConstraint
+
+joinConstraint : joinType (joinNode) [distributionType]
```

Comment:
Wondering if you considered a slightly different grammar on the constraint where you stick to a single prefix/infix/postfix notation of the join "metadata" parameters, rather than splitting the join type and distribution type on separate sides of the actual join arguments

When I read some of the examples from left to right, I feel it's initially hard for me to parse. Keeping the "metadata" together might be more intuitive to parse in my opinion. It could also very well be my lack of focus on this area which contributes to my confusion though :) 

Example

```
IJ (LOJ (a b) [R]) c
```

Reading left to right out loud (and in my head) I hear:
"Inner Join the result of a left outer join of a and b which is a replicated join with table c"

versus a slightly different notation where you might define a constraint as 
```
joinConstraint : [distributionType] joinType (joinNode) 
```

The previous example now becomes
```
IJ ([R] LOJ (a b)) c
```

"Inner Join the result of a replicated left outer join of a and b with table c"

The second one feels a little easier for me to parse when the "metadata" about the join is in one location before the inputs. However, I think you could also do a postfix or infix version. Prefix felt easier for me at first though.

Reply:
I thought about this some more, and I think the reasoning behind leaning towards the prefix notation in my examples comes from the fact than in English grammar, you will generally describe things as `<adjective> (<adjective>)* <noun>`, so I think the readability of the prefix notation is easier to parse in my head because I see the `distributionType` and `joinType` as "adjectives" which are describing the join. Granted [there are exceptions to that rule](https://www.ucl.ac.uk/internet-grammar/adjectiv/postpos.htm). I could also be convinced that a postfix notation is more suitable, as we don't have to design the constraint language to be readable in English. Just some thoughts

Reply:
- IMO the join type is a "verb" and the table names are "nouns", so `(a IJ b)` is read as `a INNER JOIN b`. The properties `R` and `P` are extra attributes of the join, so they can be skipped or specified after the "join sentence" is made

- For outer joins, it is easier to understand what is the outer-side if it follows the SQL convention, so `(a LOJ b)` is well understood as `a LEFT OUTER JOIN b`. This also makes more of a difference when nesting goes beyond a few tables, e.g consider -
```
In order : (((1 LOJ 2) ROJ 3) LOJ 4)
```
vs
```
Prefix : (LOJ (ROJ (LOJ 1 2) 3) 4)
```

Reply:
I can see how using infix makes seeing the left/right-hand sides of the join easier. In that case, should we update the `joinNode` definition to use infix notation for the join type? The current spec has `joinType` being specified before the relation name.

Reply:
I feel that we always see distribution as the last piece of a table string in EXPLAIN plans. So the current syntax looks alright for it. I'm struggling more with LOJ being prefix instead of infix. Any particular reason for going with prefix ?

Reply:
Yes, I think the grammar here is buggy. @ClarenceThreepwood can you fix up the grammar to reflect our understanding here. I think infix representation is easiest to understand, and we should start with it. The grammar can be expanded at a later stage to allow for terser specification.

Reply:
@ClarenceThreepwood  I took a stab at making the grammar more precise (and closer to parseable ANTLR). Please let me know if this is in-line with what you had in mind

Reply:
I've updated the grammar. It parses correctly in my local testing